### DISCUSSION TITLE
Regions and selections in Wave view

### DISCUSSION THREAD

**Discussion by mattyostrowski** [Ircam-RnD/catart-mubu#119] [Bug, Issues, and Suggestions]
I'm a bit unclear on the differences between regions and selections, and how to explain them to viewers. It might also be helpful in the wave view help to show the output from the imubu when you select a selection/region, and perhaps make @outputselection and @outputtimeselection visible there as well.

I also notice that in the Wave View help page/keyboard interactions pane it says 'select region', but it only selects a region _if_ you have the Region tool selected. Otherwise it selects a selection, as the default is Edit mode

**Comment by diemoschwarz** [comment]
Good point.  They seem to work slightly differently:
- **selection** is transient and set by the user for editing (delete/cut/copy), and in split view can select just one "lane"
- **region** is mainly meant for display and setting of time-ranges by the patch (e.g. display the grain extent in mubu.granular help, or set a region in a buffer to granulate from), and goes across all lanes

We should clarify that, and not conflate it into _region selection_

Hope @RiccardoBorghesi  sees this here on catart-mubu.  I think you can open mubu-related issues like this rather on https://forge-2.ircam.fr/ismm/max-dev/mubu-and-co/, Matt.

**Reply by RiccardoBorghesi** [reply]
yes you are right. "interactions pane" in Wave help should says 'data selection'. 
And showing the output of the selection is also a good idea (but should it be included in all the editor helps?)
And also, where do you think is the best place to put the explanation of the difference between selection and region?

**Reply by mattyostrowski** [reply]
I'm going to keep this thread here for the time being, and put any new issues on the other page....

@diemoschwarz:

Editing with selection seems limited:

- cut/copy is disabled in all views, and trying to cut/copy does not send the data to the Clipboard.
- delete appears disabled in the Edit menu, but works in the following views:

-  bpf
-  multibpf
- markers 
- spline
- scatterplot
- score

In none of the above is the delete undo-able.

I thought Region was mostly for display, and in that case I wonder if the Region tool is useful, or just confusing. For example, in scatterplot, the Edit tool lets me pick an arbitrary rectangle, but the Region tool is always top-to-bottom on the Y axis. 

@RiccardoBorghesi: I'll certainly cover it in the tutorial. Not to give you more work, but I notice that the one category in the main .maxhelp that doesn't have a subpatch is 'tools'. Maybe could do it there, (but if I had my way, I'd dump the region tool entirely) How about the 'Components' window?

I also notice that it imubu-region.maxhelp, there is a bunch of cut/copy/paste stuff connected to Regions which does not work, and also conflicts with what Diemo has written above....

**Reply by RiccardoBorghesi** [reply]
ok thanks, I'm going to see all of that. However I'll tell you right away that unfortunately nothing is undable in imubu! Undo is a big work for a future major release

**Reply by mattyostrowski** [reply]
Funny that I never noticed the lack of undo before! I don't really do much editing/interaction with imubu, so I never noticed....

**Reply by RiccardoBorghesi** [reply]
ok, some quick answers:
- there is no way to tie Max menus with imubu interaction. The cut/copy/paste menus will always be disabled when you edit/interact with imubu (in RUN mode). They will only be enabled when you interact with imubu as an object max any within a patch (in EDIT mode). copy/paste inside imubu works through messages (key/mouse etc.) sent by Max to objects, messages that have no relation to the menus
- copy/paste works with almost all editors, except a few where the meaning or implementation would be too complex. However, copy/paste is currently a bit broken. I'll finish a couple things and get right to it.
- Region was introduced to be able to make a selection on all tracks of a buffer at the same time. Copy&paste, if it is not broken, also works on a region (== multi-track selection). Normal selection only works on the current track, which is the foremost.
- Yes, you are right. Every tool should have its own help patch. In a near future.

**Reply by RiccardoBorghesi** [reply]
- imubu-region.maxhelp the copy/paste of a region seems to work here. Why do you say it conflicts with what Diemo said?

**Reply by mattyostrowski** [reply]
Ah! So copy/paste in mubu is private to the Mubu universe -- I didn't get that (I got started on all this trying to copy some bpf data and paste it to a textfile) -- I'll definitely mention that in the tuto.

Re your last question: have a look at Diemo's first comment after I started this thread....

**Reply by mattyostrowski** [reply]
Playing around with the copy/paste a bit, I noticed a couple of things you may already know, so sorry if this is redundant -- Here's a little movie:

https://mab.to/t/SomNj8wkKeq/us2

Executive summary:

- pasting multitrack region into prestructured mubu pastes entire audio file, and sleected parts of other tracks
- pasting wave to another mubu using cmd-v doesn't get all data, but using addtrack paste is fine
- pasting multitrack into fresh empty mubu works in the helpfile, but not in a new imubu

**Reply by RiccardoBorghesi** [reply]
Thanks Matt,
with all your reports copy&paste will have a new happy life! At the moment I have fixed almost all the bugs you showed in the video.  I add  a couple of details about mysterious bugs:
1) **(pasting multitrack region into prestructured mubu pastes entire audio file, and sleected parts of other tracks)** from  the region help. This happens because in the imubu in which you paste there is already an audio file loaded (but you cannot see it  because the view bounds are wrong). So that is one of the rare cases where the paste does what it is supposed to do.
2) **(pasting multitrack into fresh empty mubu works in the helpfile, but not in a new imubu)**  : the difference between the two cases is the layout. In the one that works the layout is juxtaposed, in the other is superposed. This because in the superposed there is an hidden component catching the keyevents and throwing them away. But is fixed now.
I still have a couple of things to do and I'll make you a new release.

**Reply by mattyostrowski** [reply]
Thanks Riccardo -- looking forward to next iteration!

**Reply by mattyostrowski** [reply]
Looking at the the last update, I see region/multitrack copy/paste working fine if layout is juxtaposed, but not if superposed, even if both imubus are in the same layout mode. Is that what it should be doing?

**Reply by RiccardoBorghesi** [reply]
Hi Matt, I tested and here region/copy/paste in superposed layout seems to work. if you can give me a patch where it doesn't work that would help.

**Reply by mattyostrowski** [reply]
It may have something to do with the order you do things in or something else weird. Anyway, here's my failed attempt.... (Damn -- github won't let me attach max patches....

<pre><code>
----------begin_max5_patcher----------
3226.3oc6bs0iaqaD94c+UPXTz9hWCQJpaAEE4zK.sOjWJ5aIEKnsns0IxRp
5xd4bv4+9Y3EIQZKIauY8ljFGj3rlTRblgeyGmgbz9q2dyrk4OwqlgdG5ina
t4Wu8lajMIZ3F82uY1N1SqRYUxKa1p7c63Y0ylq5ql+Tsr8T1y4M0nplBdYQ
dEO9cnBVUMGkjg1xK4n3bdU1epF8Xd4mau6hRdE7vX0I4Y2mljwWk2jIedj1
qfUuZaR1l6K4qpUxY.wYgybjWn6h.ZTTTXnSPjS.1gNG4FFH5iPW3f9u5GQV
ytjrTdsT9w5FShkRc9xe9Nh+r9qDzg1K0Qz3uc6shOl+5Xc94lmpYCZcRpPa
xyimMtZiCiFWs8bOe016xq1U7TP9Qk7MvDL5OhVkW7b6EcpS2XbDVpbD+Ej8
UbrmDKPwmihSt7JdQY9JdUEJodpYTmH4zlan3SspbVyg3n2.naNKFwPqSR4S
nKAX4jjqyKVUBd8TkjcMKaZebOvKWBdblCab9NVRVYSJu79kaVkmlWpmQjRu
4Gcx+iIYw.aUxuHeRNsVo77zkrx6aa2sqC1xp6AO8DA0l4cXN1s2E1S24p7r
5jr7Fff.twrMiciOjTkrL0RmXoIaxdHg+3RvmJtx7NAqYQSccxNtxebOIpjk
sgaKPcpQdYRK+r4s.BXQJaEeadZLuzZJUp3mpQMufsJo9YvaYCXaqRdvx3tp
orJu79LNC5r1RiT2mroENiiJ8oTwP5Sj9YzPoiFwOvPDzCx57zz7G2IL8V1F
dEaWQJWf9sroKaVulWtZaN.sf61TekX+9OvGNXBybSVRs8HI3HGZlUZQGnca
Q3.rTUQZRMfMqt+rjsUIkqR473M78mBG91zxcOpqeP5tFaIc.UoWVGnSE78A
VZi8Tfk.ukmrYqzf5tv1ioZUIL2JbROML45RN21IWI.el+7PCemKyHC5.Jzf
Tgrl57CccaGjsrB4iXF7HmM.gvoB.a4qNUGTywXSoht1wPhK4f4pZaIqdPZQ
SBvNX6INzqgG9tbke+PD2CQtJ8eebKmmpl.FglaeMguaI2pAoK2oZSUB0gd.
XOXMQpGlPCv9ddtjfff881zbamlqlBGd.GkwhntyNTYG..VvJY630Pm7LltO
mAW8cOePX4kL3NGhjFVoHaS950U758AIMEwZ7AlzSWWkwJZwMXGm9NrYKNU3
RKavIYIgXa3kYrzNZ8Q3T6c6f0OyyplMD0C7PROKtRKXUni7OPTtDmHRj3t7
ij+oMN2AaxxoVfVa4gHCaFGxSrm0s6laiMWltxnjsBMV4vXiOFL7BAWfzFJh
oYPC3iIw0aml5ta0MCNY.iBgNvUFxYyLrxVbRlXNwCstjs5yhNtCaKLC3nXx
2ztfs6dbbvyKqZHxn1KXfmqdDMuq6fYzWPntPLcUrM7imN9T4g3RkYM6E3KQ
rQdpvjFK5cx.QuO6XzGGLacMt9qw0+Riq2yyHrdOBUgWwWCq+RFVOkDsfD5A
rDN3HGhmGUrND0yegefeXTH0A66FfwgWi6+Zb++HG2OIzeQjOtOzeJ4BG2OF
eMv+uuB7eBE5RFNO90KbdkNcMf9W4.5U.tIO0.rS+Vsqi84b1oc5ruD1.6Io
kLQvrujX4y3OBBygGbBRPbbmbNEw.O0m+Ed4IXM7kgBhCORpKCYNnibvClCJ
uTqJZcAVlBXPfTPprcBAduhBiluw3VDFfeV4+GNuqIHs.YS9cMUxE.L08600
JC78SpA8toTwL8jeqfKW.DhWOqIQJJpFgoBsHIM5B56JH7ZMsFL2z1cuwj5n
XthjoCF4HolnQQ8VSwx3o4q9r0xlh314YIYlmlqU2w70rlzZHZ7r5NGexhA6
esVDGry10fl8SkIrzNMPrtddlPHrlJDM2Nb.RwSsoblJi7JfkiF3lAnAXWFo
yJPIap.t.Cp.RamZZhCYI.hX95Zc2EIYY6YEqyKFuyRQbti28xbnycS8rk8T
ArdpduG.E02Wwdv1ZWyRS0tl1O9mXYI6.FXQ5lpEl55TwdrsOzKrcOOLPOw.
HeU+BENlfA3xSJZAQy5lkiS1nSWrusZ1lJ6VppeVYzMZpYo1I99ZNj1mdcDv
oIisJuusNWXHi0p5ps4OVouu8h8YuxlvzG2j1yp882JCAkTm+6HY4pNWXZOa
uqwopbH4liQGCroviyqOI2twkzdDwlVW4fkEyexf5QuXflE5EZgrVHbZSjGc
3kBGk++fRBvdXFcYwWnI7UzpXsr4AEaAB7dAlP9QrZAT8AgRjKc5qJuhoraj
gsa3vwwVjiZXl2sOfWdiiXmCWzUUFhucWLjl4cXHdi3jbTUJPG8tmSxPuW3r
ij+jrwEhcHDkAV01FDqofvAXm1F1lWHxjA8d3tV.VMHb0+GxS+837Gyj61DL
Gfnn22toXpXrQvxlvmuGtobzllDzm.uI.6IVFDsCV2K4QfhdNpdaxpOmIDd2
4HYdDhk7QkPzrKyebNpeaD.DIRvcAWjbMeD9SVNqV0ciqQG8Lm+q+8e6m9vw
77zgJHi6JvUFpf4Nvc5XH53XH5TXHI9YtfJs9MEJc19YXO+EdeY9YN++heVZ
dSr.F+tJ9FwZI+Y3+iaJkwL9oOMWzJ6I8OvYY+Es6H70w7QXkeFh2deOfd+E
xbjdGcPKSgrJlijD7PKn+tdfmiza7LZQf2EwawWUcT3PEgK8E6sD9clyxow6
tdc86.nJL0MB8X6S.3Tgqcgj6aGaCjAeC7SuWdu5VAuSSjhIrXhoVxKmHT4i
qRcRWFbT2WvTazkbpsOmGgZORnqRsSz+vy4vbR4p1QUKwHWK8DBduNIqKCvO
1SeAScmFj6bEBW0y9nBA4BJCjSTFhtfxfHVrSRHBujBA8TgDWTov4DMERw04
HNLR5fQ7X51Ph8XJZ20hgEXysx.jjOHyC0NUKysJ4ixb.Mkx4FllRXUzCS5U
PQKhdrqDZ0J0slOfYhz+iuWsUG2ypqKSfrtT5p4dccVojuIMeIKcuDmGJi8a
6s3uRE8.KNVtwgxRqepMMzW8FDnHqwdQme8N37lTvCGo5N5ecBlr5NToG6ge
wU2QzWUkcUJGBXKM8bewATEI.kHSxU+dB3dFZcvWUstEKCQV.DppnfWVr97e
6I7MrBZn9YYF79pZFJ4rXVQAGRAOtrYWZddwBVxZsAQFC4DX+1HsUulENtmM
x26qqetHr2ITOu.Ip9Nccv3b1pmK4MQ8fG9xIOEGZDQoH3SQQF3PbbcN+C0h
LpRN+K4fsFatrstFP+A7TVhPEHUcbVpSf4rlRIgeSQVODpUs5KVdBPHUMGdd
pH8Z0FdsZCekq1PuiBXIpcuwScFytt9WK2vKY4F5CKWG5gcoggdAgNAXhm3.
98CWf8gH2c7HggTmPiZq5Z4FdtkaH9ajxMrKMQycHb1kqVDELj5Cx7inrlzz
4S9YOB6a3ZXD+ZVCitzfvETWO2.JlF46EE36cIpgQwRMvjw5jM5kp62Xe04d
IDO6c129asj9heFtRXYoGTTNHL0u6eNKHQgNtQhSFCf3H0dfNGM7Yscrwbfi
ey534LFiC5w7b41S.ZOHiiM7VGsgZrpRkzRVOup7Lvzv1cVROQe5HwI.WI6Y
ItC0WIdCbj4Wq3zqupYey8plcs1TuH0lpHKO65Scahc4.0mHnyw2VFWfV1qG
PFF4KBxVm9q32zGvpNXuPLE5s2fKHxTHB+dasYkCNY7OKsWXLJxgR8bBfk4.
APh+8bHDB.8ITL7WOoukKHbXpiaXnquPJMcIVByvajRk0pvfc3POUvqh35Qh
nXbfOMXvlLe1F7s9m+FbzBdfT5f3QVklnfhshXJeCa0yVBJdpsDo8LFeQISe
NPGxT6lM1KPTLEfMSkS1U3xfvkvebfK9G+WZR5T2uhVFAsfwN+3fW7NNdAq2
E1q3kQvKt+3.WnSEHSndKAoWgKS.Whv+3.WvS9xJ4QtF364x0P9wA73dBEsx
UtlogKzuugK68qgwGX80P0+jmV7eRpSO7TW+GOUjJ2ULTxGZ9qcGZ4nuxbCi
uTu75peiTpJi0.q8ztG53NIzYvIZysNDO8aboYsuINkMCqnTZrKgUkQc+RDT
aZOrz.MqLvQpJPqJJc1tj3hbX5rpqbpTUyupZRv.GtrDSr+ptxL78bEWq7ai
fFNUAO5KWv8UBipVzgDokYPu+We0E7fyVvyKiUuusNW7AFiGdjweQirmyILz
F0i7KefNCU7KZfnmv.QouBCjK4TLcguFijya0HQBeifCD5az.gcdY.OE88du
h7hAYuWM98ds3O7Uhe7WG98eU3k057XuB729a296.3W0ihB
-----------end_max5_patcher-----------
</code></pre>

**Reply by RiccardoBorghesi** [reply]
I confirm that everything works for me, also in your patch. Maybe I fixed a bug without realizing it. I'll make you a release as soon as possible and see if it works for you too. I'll let you know.

**Reply by mattyostrowski** [reply]
Interestingly enough, the above patch crashes in both 8 and 9 with the new version of MuBu. Can you confirm? Here's the crash reports:
[Max_8_crash.txt](https://github.com/user-attachments/files/17559793/Max_8_crash.txt)
[Max_9_crash.txt](https://github.com/user-attachments/files/17559794/Max_9_crash.txt)
...

I can confirm that with pre7 it opens in both. I suspect it has to do with the mubu.process objects in the 'auto-track-analyzer patch, as other patches with that subpatcher crash as well...

**Reply by diemoschwarz** [reply]
with the patch you pasted above?   can't crash it here.  It looks like an old thing that should be fixed since a while.

**Reply by mattyostrowski** [reply]
Weird -- it's going down in 8.6.5 and the 9 RC. Opens fine with older version of MuBu....

**Reply by RiccardoBorghesi** [reply]
it works for me too, no crash

**Reply by diemoschwarz** [reply]
any old pipos in your path?

**Reply by mattyostrowski** [reply]
Found a couple lurking in some MforL devices I built for Ben Neill, but cleared all pipo~s and pipos out of the search path or off the machine. Still no luck.

**Comment by diemoschwarz** [comment]
also, imubu can be made to spit out all interaction with the **selection**, also in matrix view, which is not documented.

Over to here: https://forge-2.ircam.fr/ismm/max-dev/mubu-and-co/-/issues/424
